File: Mediator/c#/Mediator.cs
interface IMediator
{
    void Notify(Employee employee, string msg);
}

abstract class Employee
{
    protected IMediator Mediator;
    
    public Employee(IMediator mediator) => Mediator = mediator;

    public void SetMediator(IMediator mediator) => Mediator = mediator;
}

class Designer : Employee
{
    public Designer(IMediator mediator = null) : base(mediator) {}

    public void ExecuteWork()
    {
        Console.WriteLine("<-the designer is at work");
        Mediator.Notify(this, "the designer designs...");
    }

    public void SetWork(bool state)
    {
        if (state)
            Console.WriteLine("<-the designer is released from work");
        else
            Console.WriteLine("<-the designer is busy");
    }
}

class Director : Employee
{
    private string _text;

    public Director(IMediator mediator = null) : base(mediator) {}

    public void GiveCommand(string text)
    {
        _text = text;
        if (_text == "")
            Console.WriteLine("->the director knows that the designer is already working");
        else
            Console.WriteLine("->the director gave the command:" + _text);
        Mediator.Notify(this, _text);
    }
}

class Controller : IMediator
{
    private Designer _designer;
    private Director _director;

    public Controller(Designer designer, Director director)
    {
        _designer = designer;
        _director = director;
        designer.SetMediator(this);
        director.SetMediator(this);
    }

    public void Notify(Employee employee, string msg)
    {
        if (employee is Director)
        {
            if (msg == "")
                _designer.SetWork(false);
            else
                _designer.SetWork(true);
        }
        else if (employee is Designer)
        {
            if (msg == "the designer designs...")
                _director.GiveCommand("");
        }
    }
}

class Program
{
    static void Main()
    {
        Designer designer = new Designer();
        Director director = new Director();

        Controller mediator = new Controller(designer, director);
        
        director.GiveCommand("design");
        
        Console.WriteLine();
        
        designer.ExecuteWork();
    }
}

File: Mediator/c++/Mediator.cpp
#include <iostream>
#include <string>

class Employee;

class Mediator
{
    public:
        virtual void notify(Employee* employee, std::string msg) = 0;
};

class Employee
{
    protected:
        Mediator* mediator;

    public:
        Employee(Mediator* med = nullptr) : mediator(med) {}
        void SetMediator(Mediator* mediator);
        virtual ~Employee() {}
};

class Designer : public Employee
{
    private:
        bool isWorking;

    public:
        Designer(Mediator* mediator = nullptr) : Employee(mediator) {}
        void executeWork();
        void SetWork(bool state);
};

class Director : public Employee
{
    private:
        std::string text;

    public:
        Director(Mediator* mediator = nullptr) : Employee(mediator) {}
        void giveCommand(std::string txt);
};

class Controller : public Mediator
{
    private:
        Designer* designer;
        Director* director;
    
    public:
        Controller(Designer* designer, Director* director);
        void notify(Employee* employee, std::string msg);
};

int main()
{
    Designer* designer = new Designer();
    Director* director = new Director();

    Controller* mediator = new Controller(designer, director);

    director->giveCommand("design");

    std::cout << std::endl;

    designer->executeWork();

    delete designer;
    delete director;
    delete mediator;
    
    return 0;
}

void Employee::SetMediator(Mediator *mediator)
{
    this->mediator = mediator;
}

void Designer::executeWork()
{
    std::cout << "<-the designer is at work\n";
    mediator->notify(this, "the designer designs...");
}

void Designer::SetWork(bool state)
{
    isWorking = state;
    if (state)
        std::cout << "<-the designer is released from work\n";
    else
        std::cout << "<-the designer is busy\n";
}

void Director::giveCommand(std::string txt)
{
    text = txt;
    if (text.empty())
        std::cout << "->the director knows that the designer is already working\n";
    else
        std::cout << "->the director gave the command: " << text << std::endl;
    mediator->notify(this, text);
}

Controller::Controller(Designer *designer, Director *director)
{
    this->director = director;
    this->designer = designer;
    designer->SetMediator(this);
    director->SetMediator(this);
}

void Controller::notify(Employee *employee, std::string msg)
{
    if (auto dir = dynamic_cast<Director*>(employee))
    {
        if (msg.empty())
            designer->SetWork(false);
        else
            designer->SetWork(true);
    }
    if (auto des = dynamic_cast<Designer*>(employee))
        if (msg == "the designer designs...")
            director->giveCommand("");
}